### ISSUE TITLE
investigate subprocesses under new input regime

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#2217]
in #2216, @kmarius brings up the fact that `system("bash")` within a Notcurses context immediately exits. i have no idea if this has always been the case, or recently changed (he might have more knowledge).

now in rendered mode this was never allowed, but it ought work in direct mode. furthermore, running an interactive bash shell ought work using an `ncsubproc()`. with that said, i've never tested any interactive subprocess of Notcurses, and i have no idea if this works, or how it fails if not.

look into it, find out what's the current state of things, and support out what we can. how do `ncsubproc`s get input? do they get input? investigate all this.

**Comment by kmarius** [comment]
Yep, this definitely worked up until 2.4.2. I ported a small application from `ncurses` to `notcurses`, more or less changing `WINDOWS` to `ncplanes`. The only issues I ever had were colors being off in the because they were not reset, and console output mangling `notcurses`.

**Comment by dankamongmen** [comment]
cool; i assume you were then running the subprocess in some area of the screen that Notcurses wasn't touching? fwiw, this can lead to problems because rendered mode assumes nothing else is touching the screen. if your bash process moves the cursor or changes the active style etc., the next time you render, things might get befouled (but they might not).

either way, the input aspect of things needs resolving. i'm not sure off the top of my head what to do here; this is kind of complicated with the new way of things. i'll have to give this a bit of thought. i don't think `NCOPTIONS_DRAIN_INPUT` would help you here, either, except for the case where input is not coming from a terminal.

this might take some thought. what you're doing is technically not something i support, but the current situation could certainly be improved. i don't know how technical you are, but feel free to throw out any ideas.

the main reasoning behind the recent input changes are that i need to be able to read terminal-generated data (cursor location reports being primary among them) without relying on the user to be actively reading (i.e. i need be able to do it in my own context). there might be arbitrary amounts of regular input in the queue ahead of what i'm looking for, so that needs be buffered. the reasoning behind not supporting other programs writing to the screen in rendered mode is that i get tremendous efficiency gains by assuming it looks and is configured like it was when i last rendered.

an arbitrary interactive subprocess works against both of these fundamental design principles. i'm thinking then that what i'd need do is provide an API that lets you say "hey notcurses, i'm gonna do some stuff over here, don't write at the moment, give me the inputs, and i'll let you know when you can go back to your stuff". it's not great for a subprocess that also has an active UI around it, but it would be something.

#1091 might fit into here a bit, and maybe this is even a headway into #381.

so yeah, #2216 i can get tonight, but solving this one well might take a bit longer, and might require some small modifications to your program. i'm sorry about that =/.

**Comment by kmarius** [comment]
>i assume you were then running the subprocess in some area of the screen that Notcurses wasn't touching?

No, I was calling `system()` on the same (main) thread as all drawing logic and replacing the host application (temporarly). Think of my application as something like `nnn` or `ranger` where you would navigate to a file and open it in `vim` to edit it and then return to navigating.

With `ncurses` the code used look look something like

```
def_prog_mode();
reset_shell_mode();
kbblocking(true);
endwin();

system("bash");

reset_prog_mode();
kbblocking(false);
refresh();
```
whereas with notcurses it used to be 
```
system("bash");
notcurses_refresh(); // maybe
notcurses_render();
```
After porting I was happy it was working so easily because this has the additional benefit that the UI is not torn down and redrawn which used to cause flickering when the command exits immediately.

> solving this one well might take a bit longer, and might require some small modifications to your program. i'm sorry about that

Don't worry about that, I will stay on 2.4.1 for now. It is an `ncurses` port after all and doesn't use the cool new features (yet).
Also it is entirely possible that I am just holding it wrong.

**Comment by dankamongmen** [comment]
> Don't worry about that, I will stay on 2.4.1 for now. It is an `ncurses` port after all and doesn't use the cool new features (yet). Also it is entirely possible that I am just holding it wrong.

nah, you're doing something i don't support, but i could support it better if it's something people really want to do. thanks very much for the reports!

**Comment by WSLUser** [comment]
From what I've read, maybe https://github.com/dankamongmen/notcurses/issues/1827 would help for the simple direct port and not taking advantage of the true features of notcurses.
